src/dense.jl

```diff
@@ -602,26 +602,35 @@ function (^)(A::AbstractMatrix{T}, p::Real) where T
     checksquare(A)
     # Quicker return if A is diagonal
     if isdiag(A)
-        TT = promote_op(^, T, typeof(p))
-        retmat = copymutable_oftype(A, TT)
-        for i in diagind(retmat, IndexStyle(retmat))
-            retmat[i] = retmat[i] ^ p
+        if T <: Real && any(<(0), diagview(A))
+            return applydiagonal(x -> complex(x)^p, A)
+        else
+            return applydiagonal(x -> x^p, A)
         end
-        return retmat
     end
 
     # For integer powers, use power_by_squaring
     isinteger(p) && return integerpow(A, p)
 
     # If possible, use diagonalization
     if ishermitian(A)
-        return (Hermitian(A)^p)
+        return _safe_parent(Hermitian(A)^p)
     end
 
     # Otherwise, use Schur decomposition
     return schurpow(A, p)
 end
 
+function _safe_parent(fA)
+    parentfA = parent(fA)
+    if isa(fA, Hermitian) || isa(fA, Symmetric{<:Real})
+        return copytri_maybe_inplace(parentfA, 'U', true)
+    elseif isa(fA, Symmetric)
+        return copytri_maybe_inplace(parentfA, 'U')
+    else
+        return parentfA
```

Comment:
Perhaps it might be better to return `fA` here, as we don't really have a notion of what the parent of an arbitrary matrix represents. In common cases, this branch will be hit when `Hermitian(A::Matrix)^p` returns a `Matrix`, so returning `fA` directly seems ok for type-stability.